c/examples/cpp_sorting_example.cpp

```diff
@@ -0,0 +1,121 @@
+#include <cstddef>
+#include <vector>
+#include <algorithm>
+#include <stdexcept>
+#include <sstream>
+#include <iostream>
+#include <type_traits>
+#include <tskit.h>
+
+static void
+handle_tskit_return_code(int code)
+{
+    if (code != 0) {
+        std::ostringstream o;
+        o << tsk_strerror(code);
+        throw std::runtime_error(o.str());
+    }
+}
+
+struct edge_plus_time {
+    double time;
+    tsk_id_t parent, child;
+    double left, right;
+};
+
+int
+sort_edges(tsk_table_sorter_t *sorter, tsk_size_t start)
+{
+    // if (sorter->tables->edges.metadata != nullptr) {
+    //    throw std::invalid_argument(
+    //        "the sorter does not currently handle edge metadata");
+    //}
+    if (start != 0) {
+        throw std::invalid_argument("the sorter requires start==0");
+    }
+
+    std::vector<edge_plus_time> temp;
+    temp.reserve(static_cast<std::size_t>(sorter->tables->edges.num_rows));
+
+    const auto *edges = &sorter->tables->edges;
+    const auto *nodes = &sorter->tables->nodes;
+
+    // NOTE: this future-proofs the type against changing to int64_t
+    // for tsk_id_t.  It will also look odd to a C programmer,
+    // but the edges->parent[0] need not be valid at run time.
+    // Rather, it is checking what the return type of the expression
+    // would be, which is tsk_id_t& (with the & being due to decltype
+    // not returning the decay type).
+    const auto num_nodes
```

Comment:
Maybe I'm being dense here, but why not use ``tsk_size_t num_nodes = nodes->num_rows``? Most of the point of the typedef is that allows us to change to 64 bit later, if we want. (The rest of the point is readability, but some would argue against that and say it's better to be explicit about what the types actually are.)

Likewise for the other types - I think it'd be more readable to use tsk_size_t than type inference.

Reply:
Typically, one would prefer to just use `auto` here, following the [AAA](https://herbsutter.com/2013/06/13/gotw-94-special-edition-aaa-style-almost-always-auto/) rule of thumb ("code for the interface, not the implementation"). However, that leads to cast warnings later, but those would go away if the node check during copy disappears.